### PR TITLE
Bluetooth: Audio: Volume Control Service and client

### PR DESCRIPTION
**Pull request by Thalley**
This PR implements the Volume Control Service (VCS) server and client. 

The implementation supports and uses the Audio Input Control Service (AICS) server and client implementations as well as the Volume Offset Control Service (VOCS) server and client implementations. 

Besides the implementation, a full test suite that exercises the entire API is also added using Babblesim.

### CODE REVIEW

**Comment by Thalley** [comment]
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/34169 and https://github.com/zephyrproject-rtos/zephyr/pull/34127 to pass CI

File: tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh
Comment (aescolar) [inline tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh:27]: Out of curiosity more than anything:
Why do you have 2 tests in the same script? (having them as 2 separate scripts allows running them in parallel, and would save you the need for the verbosity)

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh:27]: Mostly due to my local development process I suppose: Easier to run just a single script to test the entire VCS. 

I could split it into two scripts but then I'd probably add a third script to run both of those anyhow :D

Comment (aescolar) [reply tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh:27]: aham :)
Maybe you'd want to use as 3rd script the one that is used in CI:
RESULTS_FILE=${ZEPHYR_BASE}/myresults.xml SEARCH_PATH=tests/bluetooth/bsim_bt/bsim_test_audio/ ./run_parallel.sh

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/test_scripts/vcs.sh:27]: If I'm just testing e.g. VCS locally, then I have no intention of building all and running all tests (which does take quite a while compared), so I probably wouldn't do that :D

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:177]: Insert "be".
Before other actions can be completed, or before other actions can be started?

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:174]: Client side only

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:188]: Server side only.

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:304]: "read" -> "set"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:318]: "read" -> "set"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:339]: "read" -> "set"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:342]: "The value"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:353]: Insert "to" - "to deactivate"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:350]: "Deactivates" -> "Deactivate", to be consistent with the imperative form of the other functions.

"a" -> "an"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:362]: "Activate an"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:365]: "to"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:422]: Documentation says "mute", function says "unmute"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:432]: Documenation says "unmute", function says "mute"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:438]: "read" -> "set"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:448]: Ditto

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:457]: Ditto

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:480]: Ditto

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [inline include/bluetooth/audio/vcs.h:482]: Is the description copied here as well?

**Review by asbjornsabo — COMMENTED** [review]
Comments so far.
I think the API is good, but that it must be made very clear how this works on server and client sides.  Also, some functions are one side only, and should be marked as such.

In addition, a lot of nitpicking.

File: subsys/bluetooth/audio/Kconfig.vcs
Comment (Thalley) [reply subsys/bluetooth/audio/Kconfig.vcs:38]: Correct, or even other vendor specific services. I don't think there's any way we can at compile time determine if the values are correct.

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:422]: Hah, I think we had that same comment for AICS, right? 
Will fix of course :)

File: subsys/bluetooth/shell/CMakeLists.txt
Comment (Thalley) [reply subsys/bluetooth/shell/CMakeLists.txt:31]: A quick glace of the CMakelists makes the norm 
```
zephyr_library_sources_ifdef(CONFIG_XXX, YYY.c)
```
So the shell Cmakelists is the odd one out (there are a few other files with similar formatting though.

File: subsys/bluetooth/audio/vcs_client.c
Comment (asbjornsabo) [inline subsys/bluetooth/audio/vcs_client.c:670]: "allocated" -> "allocate"

File: subsys/bluetooth/audio/vcs_client.c
Comment (asbjornsabo) [inline subsys/bluetooth/audio/vcs_client.c:687]: 1) "allocated" -> "allocate"
2) "VOCS" -> "AICS"

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:113]: @asbjornsabo Any suggestions on how to? 

One suggestion I have, is to refactor the callback struct to be similar to `struct bt_conn_cb` where the callbacks are defined inside the struct, and "special" cases are guarded by ifdefs, so e.g. the vcs callback structure would be something like

```
struct bt_vcs_cb {
	void (*state_changed_cb)(struct bt_conn *conn, int err, uint8_t volume, uint8_t mute);
#if defined(CONFIG_BT_VCS_CLIENT)
	void (*discover_cb)(struct bt_conn *conn, int err, uint8_t vocs_count, uint8_t aics_count);
#endif /* CONFIG_BT_VCS_CLIENT */
};
```

File: subsys/bluetooth/audio/vcs_client.c
Comment (asbjornsabo) [inline subsys/bluetooth/audio/vcs_client.c:746]: I needed to think on this one.
The VOCS discover callback is overwritten (or set) in all cases - whether it was set before or not.
But the _warning_ is only given if it is set before.

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:121]: Added more text for the conn pointer. Please re-review.

**Review by asbjornsabo — COMMENTED** [review]
Additional comments

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:119]: Rewrote the text somewhat similar to your suggestion. Tried to be as concise as possibly. Please re-review.

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:177]: changed complete to `executed`

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:342]: Remove that last part as the value is no longer copied (at least not by VCS/VOCS). I assume that, unless explicitely stated, that values does not need to remain static after the call (e.g. how it's defined for gatt_subscribe etc.)

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:482]: See https://github.com/zephyrproject-rtos/zephyr/pull/34314#discussion_r617449778

File: tests/bluetooth/bsim_bt/bsim_test_audio/prj.conf
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/prj.conf:21]: CONFIG_BT_AUDIO_UNICAST isn't will remove

File: subsys/bluetooth/audio/vcs_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/vcs_client.c:112]: I did consider adding an __ASSERT for this, but no:
The connection indexes are static and vcs_insts are static as well, so regardless of the conn (unless conn is NULL), then vcs_inst will always be non-null.

File: subsys/bluetooth/audio/vcs_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/vcs_client.c:746]: I thought that made sense; if it was not set, then it was not really overwritten :)

Comment (asbjornsabo) [reply subsys/bluetooth/audio/vcs_client.c:756]: Agreed - it is absolutely correct.  But I somehow came from a different angle.
Maybe it could be augmented with an 
else
        BT_DEBUG("Set the discover callback of VOCS")?
Or possibly a comment of "/* Set VOCS callback - overwrite if already set */ ?

Comment (Thalley) [reply subsys/bluetooth/audio/vcs_client.c:756]: Actually, the better solution here might be to simply not allow the application to set the callback (can be verified in the register function), assuming that the application does not change the callback structure after registering (but that assumption goes for basically all callbacks in Zephyr). 

@asbjornsabo Do you think that makes more sense?

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (asbjornsabo) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:112]: Set terminating zero?

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (asbjornsabo) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:197]: Termination?

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (asbjornsabo) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:680]: VCS muted

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_test.c
Comment (asbjornsabo) [inline tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_test.c:115]: In the client test, there is a size check.
(Not sure how important that is here.)

**Review by asbjornsabo — COMMENTED** [review]
Last part of first round - mostly skimmed quickly through the shell commands, though.

File: subsys/bluetooth/shell/CMakeLists.txt
Comment (asbjornsabo) [reply subsys/bluetooth/shell/CMakeLists.txt:31]: OK - suggest to not fix that inconsistency here.

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [reply include/bluetooth/audio/vcs.h:113]: I had a more naive thought - just mention it in the Doxygen documentation.
Since the rule for most functions is that they apply on both sides, it should be sufficient to note the exceptions, e.g. like 
" * (Client side only)"

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [reply include/bluetooth/audio/vcs.h]: Agree

File: include/bluetooth/audio/vcs.h
Comment (asbjornsabo) [reply include/bluetooth/audio/vcs.h]: It was familiar, yes.

File: subsys/bluetooth/audio/vcs_client.c
Comment (asbjornsabo) [reply subsys/bluetooth/audio/vcs_client.c:113]: What happens if the devices is connected to devices A, B and C, with connection indexes 0, 1 and 2, but only A and C has VCS services?

File: include/bluetooth/audio/vcs.h
Comment (Thalley) [reply include/bluetooth/audio/vcs.h:113]: I've added that for all server or client specific functions and callbacks.

File: subsys/bluetooth/audio/vcs_client.c
Comment (Thalley) [reply subsys/bluetooth/audio/vcs_client.c:113]: Then 1) this (notify) callback will never be called for the connection between the device and B (since it never subscribed), and even if that happens (i.e. the server notifies with a wrong handle or something similar to that), then the `vcs_inst` pointer is still a valid pointer, but where the callbacks are simply not registered.

File: subsys/bluetooth/audio/vcs.c
Comment (Thalley) [reply subsys/bluetooth/audio/vcs.c:397]: The `#if defined(CONFIG_BT_VCS) || defined(CONFIG_BT_VCS_CLIENT)` is superfluous as vcs.c is only compiled if one of those are true anyhow. 

I'm not sure what else you mean by your comment.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:112]: `strncpy` should handle that for us.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:197]: `strncpy` should handle that for us.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_test.c
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_test.c:115]: Changed to use `strncpy`, but should not be necessary as we use the Kconfig that will have truncated the value from the service implementation anyhow :)

File: subsys/bluetooth/shell/vcs.c
Comment (Thalley) [reply subsys/bluetooth/shell/vcs.c:217]: Yes, I'm not sure how much we want to care about this to be honest. It doesn't seem to be done "properly" for most cases, but the return value is even a `long` and not an `int`. 

So yeah, ideally we would take all the results an store in a `long` value, check the value range and then use the value.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c
Comment (asbjornsabo) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/vcs_client_test.c:112]: "The  strncpy()  function  is  similar,  except that at most n bytes of src are copied.  Warning: If there is no null byte among the first n bytes of src, the string placed in dest will not be null-terminated."
(This is from the Linux manpage, but I expect zephyr is the same.)

The code checks against "sizeof() - 1", and also copies that many bytes.  I expected the last byte (the -1) to be used for the termination (in case of truncation), but it isn't.

File: subsys/bluetooth/audio/vcs_client.c
Comment (asbjornsabo) [reply subsys/bluetooth/audio/vcs_client.c:756]: I am starting to feel that I am making a lot of noise about a little thing ...
The thing works, and is correct, as it is, so although I was confused there is no pressing need for a change.
But I can not see now why the application would want to supply it's own callback, so your proposal makes sense.
Do as you think is best.

File: subsys/bluetooth/audio/vcs.c
Comment (asbjornsabo) [reply subsys/bluetooth/audio/vcs.c:401]: The #if that is removed, and the remaining `#if defined(CONFIG_BT_VCS)` seemed to imply that this function could also be called for the client. But I don't think it is?

File: subsys/bluetooth/shell/vcs.c
Comment (asbjornsabo) [reply subsys/bluetooth/shell/vcs.c:217]: I think we do not care.

**Review by asbjornsabo — APPROVED** [review]
All my comments have been resolved to my satisfaction.

**Comment by carlescufi** [comment]
@Vudentz please review before Friday so we can get this into Zephyr 2.6.

**Comment by Thalley** [comment]
@carlescufi We also need to resolve https://github.com/zephyrproject-rtos/zephyr/issues/34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).

**Comment by jhedberg** [comment]
I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

**Comment by asbjornsabo** [comment]
> I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

For what it is worth, before upstreaming the code has been developed and reviewed in smaller portions, as a number of changes over time, in the zephyr-bt-sig tree.

**Comment by Thalley** [comment]
> I don't mind so big changes (6000+ lines added) being in a single pull request, but you could IMO do a bit better job at splitting things up into smaller self-contained commits. E.g. the service and client could probably be in separate commits. I only took a quick look at the code itself (too much for a thorough review) and that looks fairly alright to me.

Yes, Luiz also gave that comment. Splitting the server and client is somewhat difficult as they share the same API, i.e. the vcs.h header file would be hard to split up into two commits. The alternative is just add e.g. the server and vcs.h (with references to the client) in one commit, and then the client in another, but that would be fairly weird. Perhaps I could also add the server and client in two commits, and then add the public header file. 

What I've done for the MICS PR is that I've split it into 3: 
1) Implementation
2) Shell
3) Tests

**Comment by carlescufi** [comment]
> @carlescufi We also need to resolve #34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).

We have been overriding that in other PRs, and we can do that here too.

**Review by jhedberg — APPROVED** [review]

**Comment by Thalley** [comment]
> > @carlescufi We also need to resolve #34823 before this can be merged, as the MISRA scripts are blocking this PR (and others).
> 
> We have been overriding that in other PRs, and we can do that here too.

@carlescufi I think this is good to merge now. Can you override the MISRA script?

**Pull request merged**
